### PR TITLE
Generalize traverse

### PR DESCRIPTION
**Pull request by treeowl**
We can convert `LogicT` to a naive (and naively `Traversable`)
representation and back.

### CODE REVIEW

**Comment by treeowl** [comment]
I haven't yet checked that this compiles with old GHC.

**Comment by Bodigrim** [comment]
Could you please make another push to trigger CI?

**Comment by treeowl** [comment]
@Bodigrim, it seems to be fixed now.

File: Control/Monad/Logic.hs [hunk 1]
```diff
@@ -23,6 +23,11 @@
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE RankNTypes            #-}
 {-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE DeriveTraversable     #-}
+#if __GLASGOW_HASKELL__ < 710
```
Comment (Bodigrim) [inline Control/Monad/Logic.hs:27]: It does not hurt to enable `DeriveFoldable` / `DeriveFunctor` unconditionally. Also please sort language extensions alphabetically.

File: Control/Monad/Logic.hs [hunk 2]
```diff
@@ -318,10 +323,49 @@ instance (Applicative m, F.Foldable m) => F.Foldable (LogicT m) where
...
+   -
+   - Inlining the inner traverse gives
+   -
+   -   traverse f (ConsML x (ML m)) = liftA2 ConsML (f x) (ML <$> traverse (traverse f) m)
+   -
+   - revealing fmap under liftA2. We fuse those into a single application of liftA2,
+   - in case fmap isn't free.
+  -}
+
+toML :: Monad m => LogicT m a -> ML m a
```
Comment (Bodigrim) [inline Control/Monad/Logic.hs:350]: Isn't `Applicative m` enough here?

**Review by Bodigrim — COMMENTED** [review]
This is great stuff! I have a couple of minor suggestions.

File: Control/Monad/Logic.hs (hunk 1 above)
Comment (treeowl) [reply Control/Monad/Logic.hs:27]: Sure.

File: Control/Monad/Logic.hs (hunk 2 above)
Comment (treeowl) [reply Control/Monad/Logic.hs:350]: Yeah. The thing is that it'll be awkward for people using ancient GHC and I'm not sure there's any practical benefit. It's up to you though.

Comment (Bodigrim) [reply Control/Monad/Logic.hs:350]: Let's use `Applicative`.

Comment (treeowl) [reply Control/Monad/Logic.hs:350]: No problem.

File: Control/Monad/Logic.hs (hunk 1 above)
Comment (treeowl) [reply Control/Monad/Logic.hs:27]: BTW: the (only) reason to enable them conditionally is that if the minimal GHC version ever rises to 7.10, it'll be easier to remove compat cruft if it's guarded by CPP. If that's not important to you, I can certainly enable those unconditionally.

Comment (Bodigrim) [reply Control/Monad/Logic.hs:27]: Please enable them unconditionally.

**Comment by Bodigrim** [comment]
Thanks!

**Pull request merged**